### PR TITLE
feat: add `fs/append-file`

### PR DESCRIPTION
**Pull request by HRIDYANSHU054**
adds file system append file utility to Stdlib

## Description

> What is the purpose of this pull request?
This pull request brings the node js file system append-file utility to Stdlib

This pull request:

is aligned with the purpose of achievinng feature parity with Node.js fs package. It Brings file system's powerful append file utility to Stdlib expanding its existing fs utilities.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Implementation details

-- async version
```
var join = require("path").join;

var fpath = join(__dirname, "examples", "fixtures", "file.txt");

appendFile(fpath, "beep boop\n", onAppend);

function onAppend(error) {
  if (error) {
    throw error;
  }
}
```

-- sync version
```
var appendFile = require( '@stdlib/fs/append-file' );

// Explicitly handle the error...
var err = appendFile( './beep/boop.txt', 'data to append\n' );
if ( err instanceof Error ) {
     throw err;
}
```

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by kgryte** [comment]
Thanks, @HRIDYANSHU054, for working on this. This PR is currently failing CI. Those failures need to be addressed before the PR can be reviewed.

**Comment by HRIDYANSHU054** [comment]
Hi @kgryte , Thanks for checking this PR. I m currently working on solving these issues. Majority of them have arised because the fixtures directory was somehow not pushed to github.

**Comment by HRIDYANSHU054** [comment]
@kgryte can you please have a look, I have made changes to tackle the previous PR issues. Though this might seem a silly question but why is it stuck in "Expected — Waiting for status to be reported" , is there any workaround I could do to get it cleared

**Comment by HRIDYANSHU054** [comment]
Corrected some more things. @kgryte can you once again check this

**Comment by HRIDYANSHU054** [comment]
@kgryte can you please approve the remaining workflows. I am interested in this and would like to tackle any other issue that arises or any change that is required.

**Comment by HRIDYANSHU054** [comment]
@kgryte all the issues that have been highlighted thus far during the CI checks have been solved. Can you again approve the remaining workflows so that if there are any other errors then they could also be solved and this pr can thus be reviewed. On a sidenote really thanks for your patience✨.

**Comment by HRIDYANSHU054** [comment]
@kgryte can you please have a look

**Comment by HRIDYANSHU054** [comment]
@kgryte please check

**Comment by HRIDYANSHU054** [comment]
@kgryte please rerun the CI

**Comment by HRIDYANSHU054** [comment]
@kgryte please can you again check it

**Comment by HRIDYANSHU054** [comment]
@kgryte please check these changes

**Comment by HRIDYANSHU054** [comment]
@kgryte please have a look , a Lot of CI check issues have been solved and now only the lint check is remaining to be solved

**Comment by HRIDYANSHU054** [comment]
@kgryte please check

**Comment by HRIDYANSHU054** [comment]
Finally, this has passed all checks. @kgryte please can you review it now.

**Comment by HRIDYANSHU054** [comment]
@kgryte this is now clear for review. Kindly check

**Comment by HRIDYANSHU054** [comment]
@kgryte please review my pr

**Comment by HRIDYANSHU054** [comment]
@kgryte the errors that were coming during the CI checks have all been solved. Please review this PR and tell me if any change is needed.

**Comment by kgryte** [comment]
@HRIDYANSHU054 
Would you mind checking the box in the OP indicating that you've read the contributing guidelines? This needs to be checked in order to confirm that you agree to licensing terms and the Developer's Certificate of Origin.

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/README.md]: I suggest basing this note much closer to https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/fs/read-file#notes. One thing to note is that we do not disable ESLint rules within the examples themselves. To disable lint rules, you need to do `<!-- eslint-disable node/no-sync -->` before the codeblock.

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/README.md]: Why are we wrapping in a try/catch when we know the function won't throw errors?

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/README.md]:
```suggestion
```

As stated in the comment, don't manually edit this section.

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/README.md:231]: Separate with line breaks as we do everywhere else in the project.

File: lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js]:
```suggestion
	}
	b.pass( 'benchmark finished' );
```

The benchmark should state that it finished regardless of whether an error happened.

File: lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js]: Wait...are you repeatedly appending the file over and over again? If I run this benchmark multiple times, this could lead to a huge file. You need to create a fresh `tmp` file for each invocation of `bench`, but before `t.tic()` and you need to delete that tmp file after `t.toc()`. Otherwise, this could lead to a massive amount of used resources.

File: lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js]: Same comment.

File: lib/node_modules/@stdlib/fs/append-file/bin/cli
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/bin/cli]: Use relative paths: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/fs/read-file/bin/cli#L48.

File: lib/node_modules/@stdlib/fs/append-file/bin/cli
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/bin/cli]: Spacing.

File: lib/node_modules/@stdlib/fs/append-file/docs/repl.txt
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/repl.txt]: Why would the encoding option be ignored?

File: lib/node_modules/@stdlib/fs/append-file/docs/repl.txt
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/repl.txt]: Same comment.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: Same comment.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]:
```suggestion
	* @param options - options (if a string, the value is the encoding)
```

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]:
```suggestion
	* @param options - options (if a string, the value is the encoding)
```

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: Why are you wrapping?

Also, within JSDoc comments, you need to space indent, not TAB indent.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: It should return an `Error` or `null`.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: Same comment.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts:131]:
```suggestion

// EXPORTS //
```

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: Don't do this. Nowhere in the codebase do we split arguments over separate lines.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: Same comment.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: Same comment here and everywhere else.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts:122]: @Planeshifter For some reason, linting on CI did not pick up this failure. The `index.d.ts` is incorrectly typed, while this assertion is correctly typed. I would have expected a CI failure.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: Same thing. We don't split over lines like this in test.ts files. As the saying goes, "when in Rome..."

File: lib/node_modules/@stdlib/fs/append-file/examples/index.js
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/examples/index.js]: Same comments as README.

File: lib/node_modules/@stdlib/fs/append-file/lib/sync.js
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/lib/sync.js]:
```suggestion
```

Never leave console statements lying around in implementations.

File: lib/node_modules/@stdlib/fs/append-file/lib/sync.js
Comment (kgryte) [inline lib/node_modules/@stdlib/fs/append-file/lib/sync.js]: Why are you throwing? One of the specific rationales for our `fs` packages is the ability to explicitly handle errors by returning as return values. See https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/fs/read-file/lib/sync.js#L53.

**Review by kgryte — CHANGES_REQUESTED** [review]
@HRIDYANSHU054 I've left a number of initial comments. This implementation needs a fair amount of clean-up before it can be ready for merge.

**Comment by HRIDYANSHU054** [comment]
@kgryte thanks for going through this and reviewing it. I don't why the splitting has happened in only some functions and not others even though I have configured my ESlint to highlight these issues if they occur. Also can I address this after 5th april as I am currenlty having fromm today my mid semester examinations.

File: lib/node_modules/@stdlib/fs/append-file/lib/sync.js
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/lib/sync.js]: since, this was implemented a while back, I took the implementation notes from Node fs documentaion, this is even visible in the first commit message 
The error handling approach is also different from stdlib's approach
```
-- sync version
var appendFile = require( '@stdlib/fs/append-file' );

// Explicitly handle the error...
try{
appendFile.sync( '/path/to/file.txt', 'beep boop\n' );
} catch(err) {
throw err
}
```

in accordance to stdlib
```
var err = appendFile.sync( '/path/to/file.txt', 'beep boop\n' );
if (err) {
throw err;
}
```
now when I ve seen and have gone through some of the stdlib implementations YES this needs to return the error and n ot throw it. and I will be working on this as soon as possible.

File: lib/node_modules/@stdlib/fs/append-file/docs/repl.txt
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/repl.txt]: yes, there is no need at all for this comment "The encoding option is ignored if the data argument is a buffer" in append-file

File: lib/node_modules/@stdlib/fs/append-file/lib/sync.js
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/lib/sync.js]: fixed, now in line with the new sync implementation

File: lib/node_modules/@stdlib/fs/append-file/examples/index.js
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/examples/index.js]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: fixed, this splitting was a result of the trailing comma after the last property in the object.

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts:131]: done

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: fixed, now in line with the new sync implementation

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/index.d.ts]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/repl.txt
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/repl.txt]: fixed

File: lib/node_modules/@stdlib/fs/append-file/docs/repl.txt
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/docs/repl.txt]: fixed, by removing the irrelevant comment

File: lib/node_modules/@stdlib/fs/append-file/bin/cli
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/bin/cli]: fixed

File: lib/node_modules/@stdlib/fs/append-file/bin/cli
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/bin/cli]: relative paths added

File: lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js]: fixed, but implementation strongly needs review

File: lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js]: fixed

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/README.md:231]: fixed

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/README.md]: fixed

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/README.md]: fixed, now this should be pre populated

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/README.md]: fixed, now in line with the new sync implementation

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (HRIDYANSHU054) [reply lib/node_modules/@stdlib/fs/append-file/README.md]: Noted and fixed

**Comment by HRIDYANSHU054** [comment]
@kgryte the changes you highlighted have been implemented, certain things like the new benchmarking approach strongly needs to be reviewed

File: lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js
Comment (HRIDYANSHU054) [inline lib/node_modules/@stdlib/fs/append-file/benchmark/benchmark.js:55]: uses capability of append-file to create the file if the file does not exists yet, and then deletes the file after its done to avoid wasting memory resources

**Comment by HRIDYANSHU054** [comment]
@kgryte can you approve these remaining workflows and then review the changes

**Comment by HRIDYANSHU054** [comment]
Hi ✋ @kgryte just hoping you could further review these changes, and anything more that you want me to change or add I will work on that also.

**Comment by HRIDYANSHU054** [comment]
hi @kgryte could you review this

**Comment by HRIDYANSHU054** [comment]
@kgryte could you review this PR

File: lib/node_modules/@stdlib/fs/append-file/test/fixtures/stdin_error.js.txt
Comment (Planeshifter) [inline lib/node_modules/@stdlib/fs/append-file/test/fixtures/stdin_error.js.txt]:
```suggestion
* Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/fs/append-file/test/fixtures/write_error.js.txt
Comment (Planeshifter) [inline lib/node_modules/@stdlib/fs/append-file/test/fixtures/write_error.js.txt]:
```suggestion
* Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/fs/append-file/package.json
Comment (Planeshifter) [inline lib/node_modules/@stdlib/fs/append-file/package.json]:
```suggestion
  "description": "Append data to a file.",
```

File: lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts
Comment (Planeshifter) [reply lib/node_modules/@stdlib/fs/append-file/docs/types/test.ts:122]: @kgryte Typing in `index.d.ts` matches the assertion here now, so I consider this resolved.

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (Planeshifter) [inline lib/node_modules/@stdlib/fs/append-file/README.md:144]:
```suggestion
// returns null
```

**Review by Planeshifter — APPROVED** [review]
Apologies for the late review; thanks for your PR, which looks very good now. I will merge shortly after CI has passed again after some tiny changes.

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (Planeshifter) [inline lib/node_modules/@stdlib/fs/append-file/README.md]:
```suggestion
var join = require( 'path' ).join;
var appendFile = require( '@stdlib/fs/append-file' );

var fpath = join( __dirname, 'examples', 'fixtures', 'file.txt' );

// Synchronously append data to a file:

var error = appendFile.sync( fpath, 'beep boop\n', 'utf8' );
// Function successfully executes and returns null

console.log( error instanceof Error );
// => false

// Asynchronously append data to a file:

appendFile( fpath, 'beep boop\n', onAppend );

function onAppend( error ) {
    if ( error ) {
        console.error( 'Error: %s', error.message );
    }
    console.log( 'Success!!!' );
}
```

File: lib/node_modules/@stdlib/fs/append-file/README.md
Comment (Planeshifter) [inline lib/node_modules/@stdlib/fs/append-file/README.md]:
```suggestion
// Synchronously append data to a file:

var error = appendFile.sync( fpath, 'beep boop\n', 'utf8' );
// Function successfully executes and returns null

console.log( error instanceof Error );
// => false

// Asynchronously append data to a file:

appendFile( fpath, 'beep boop\n', onAppend );

function onAppend( error ) {
    if ( error ) {
        console.error( 'Error: %s', error.message );
    }
    console.log( 'Success!!!' );
}
```

**Comment by HRIDYANSHU054** [comment]
Thanks @Planeshifter looking forward to learning and contribute more to stdlib node's environment

**Pull request merged**
